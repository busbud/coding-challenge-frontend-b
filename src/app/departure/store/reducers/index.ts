import { ActionReducerMap, createFeatureSelector } from '@ngrx/store';

import * as fromDepartures from '../reducers/departure.reducers';

export interface DeparturesState {
  departures: fromDepartures.DepartureState;
}

export const reducers: ActionReducerMap<DeparturesState> = {
  departures: fromDepartures.DepartureReducer,
};

export const getDeparturesState = createFeatureSelector<DeparturesState>('departures');
