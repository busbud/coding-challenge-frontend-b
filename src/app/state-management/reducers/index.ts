import { InjectionToken } from '@angular/core';
import {
    ActionReducerMap,
    ActionReducer,
    MetaReducer,
    Action
} from '@ngrx/store';
import { environment } from '@env';
import { storeFreeze } from 'ngrx-store-freeze';
import * as fromRouter from '@ngrx/router-store';

export interface State {
    router: fromRouter.RouterReducerState<any>;
}

export const ROOT_REDUCERS = new InjectionToken<
    ActionReducerMap<State, Action>
    >('Root reducers token', {
    factory: () => ({
        router: fromRouter.routerReducer,
    }),
});

export function logger(reducer: ActionReducer<State>): ActionReducer<State> {
    return (state, action) => {
        const result = reducer(state, action);

        console.groupCollapsed(action.type);
        console.log('prev state', state);
        console.log('action', action);
        console.log('next state', result);
        console.groupEnd();

        return result;
    };
}

const metaReducersList = [logger, storeFreeze];

export const metaReducers: MetaReducer<State>[] = !environment.production ? metaReducersList : [];
