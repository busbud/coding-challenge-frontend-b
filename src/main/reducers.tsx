import { combineReducers } from "redux";
import searchReducers from "../components/search/slice";

export default combineReducers({
  search: searchReducers
});
