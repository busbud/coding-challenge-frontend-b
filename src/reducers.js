import {
	RECEIVE_DEPARTURES,
	RECEIVE_ERROR,
	REQUEST_DEPARTURES,
	UPDATE_SEARCH
} from './types'

const initialState = {
	departures: [],
	from: 'New York',
	to: 'Montreal',
	currency: 'CAD',
	date: new Date(2018, 7, 2)
}

const reducer = (state = initialState, action) => {
	switch (action.type) {
		case UPDATE_SEARCH:
			return {
				...state,
				[action.field]: action.value
			}
		case REQUEST_DEPARTURES:
			return {
				...state,
				isFetching: true,
				isError: false,
				requestedAt: action.requestedAt
			}
		case RECEIVE_ERROR:
			return {
				...state,
				isFetching: false,
				isError: true,
				error: action.reason
			}
		case RECEIVE_DEPARTURES:
			return {
				...state,
				isFetching: action.isPartial,
				lastUpdate: action.receivedAt,
				departures: action.departures
			}
		default:
			return state
	}
}

export default reducer
