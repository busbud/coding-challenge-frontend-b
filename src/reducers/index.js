import { routerReducer } from 'react-router-redux';
import { combineReducers } from 'redux';
import composeState from './composeState';
import getStateDiff from './getStateDiff';

export const defaultState = {
	safe: false,
	date: '2017-09-10',
	tickets: [{
		id: 'Zjc5ZDhjODI6NzA4YWRkZjk',
		price: 9543,
		currency: 'USD',
		url: 'https://www.busbud.com/en/deeplink/dr5reg/f25dvk/Zjc5ZDhjODI6NzA4YWRkZjk?outbound_date=2017-09-06&return_date&adults=1&children=0&seniors=0&child_ages=&senior_ages=&discount_code&utm_source={utm_source}&utm_campaign={utm_campaign}&utm_medium={utm_medium}',
		departure: '2017-09-06T07:00:00',
		arrival: '2017-09-06T22:35:00',
		duration: 935,
		origin: {
			id: 1942,
			name: 'NY',
		},
		destination: {
			id: 1938,
			name: 'MR',
		},
		operator_id: '0e753dbf-a9de-4339-8a00-bbb6f4813d18',
		operator_logo: 'https://busbud.imgix.net/operator-logos/adirondack-trailways.png?h={height}&w={width}&auto=format&fit=fill&bg=0FFF',
		operator_url: 'https://trailwaysny.com',
		operator_name: 'Adirondack Trailways',
	}],
	lang: 'en',
	currency: 'USD',
};

const reducers = combineReducers({
	routing: routerReducer,
	app: (state = defaultState, action = {}) => {
		const merge = composeState(state);
		const stateDiff = getStateDiff(action);
		return merge(stateDiff);
	},
});

export default reducers;
