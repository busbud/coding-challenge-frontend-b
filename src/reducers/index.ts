import { StoreState } from "../types/StoreState";

const initialState = {
  language: "en"
};

export default (state: StoreState = initialState, action: any): StoreState => {
  switch (action.type) {
    case "UPDATE_LANGUAGE":
      return { ...state, language: action.language };
    default:
      return state;
  }
};
