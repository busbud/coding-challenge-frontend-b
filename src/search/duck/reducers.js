import types from "./types";

const initialState = {
    results: [],
    isSearching: false,
    error: false
};

const searchReducer = (state = initialState, action) => {
    switch (action.type) {
        case types.GET_SEARCH: 
            return {
                ...state,
                isSearching: true,
            };

        case types.GET_SEARCH_SUCCESS: 
            return {
                ...state,
                results: action.results,
                isSearching: false,
            };

        case types.GET_SEARCH_FAILURE:
            return {
                ...state,
                error: action.error,
                isSearching: false
            };

        default: 
            return state;
    }
}

export default searchReducer;
