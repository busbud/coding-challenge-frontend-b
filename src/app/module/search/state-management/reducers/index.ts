import {
    Action,
    combineReducers,
    createFeatureSelector,
    createSelector
} from '@ngrx/store';
import * as fromRoot from '../../../../state-management/reducers';
import * as fromDepartures from './departures.reducer';
import { departuresFeatureKey } from './departures.reducer';
import * as fromSearchForm from './search-form.reducer';
import { searchFormFeatureKey } from './search-form.reducer';

export const searchModuleFeatureKey = 'search';

export interface SearchState {
    loading: boolean;
    [fromSearchForm.searchFormFeatureKey]: fromSearchForm.State;
    [fromDepartures.departuresFeatureKey]: fromDepartures.State;
}

export interface State extends fromRoot.State {
    [searchModuleFeatureKey]: SearchState;
}

export function reducers(state: SearchState | undefined, action: Action) {
    return combineReducers({
        [fromSearchForm.searchFormFeatureKey]: fromSearchForm.reducer,
        [fromDepartures.departuresFeatureKey]: fromDepartures.reducer
    })(state, action);
}

export const selectSearchModuleState = createFeatureSelector<State, SearchState>(
    searchModuleFeatureKey
);

export const selectSearchFormState = createSelector(
    selectSearchModuleState,
    state => state[searchFormFeatureKey]
);

export const selectDepartureEntitiesState = createSelector(
    selectSearchModuleState,
    state => state[departuresFeatureKey]
);

export const selectSelectedDepartureId = createSelector(
    selectDepartureEntitiesState,
    fromDepartures.selectId
);

export const {
    selectIds: selectDepartureIds,
    selectEntities: selectDepartureEntities,
    selectAll: selectAllDepartures,
    selectTotal: selectTotalDepartures,
} = fromDepartures.adapter.getSelectors(selectDepartureEntitiesState);

export const selectSelectedDeparture = createSelector(
    selectDepartureEntities,
    selectSelectedDepartureId,
    (entities, selectedId) => {
        return selectedId && entities[selectedId];
    }
);

export const getSearchFormSubmittedState = createSelector(
    selectSearchFormState,
    fromSearchForm.getSubmitted
);
export const getSearchFormDataState = createSelector(
    selectSearchFormState,
    fromSearchForm.getData
);
export const getSearchLoadingState = createSelector(
    selectDepartureEntitiesState,
    fromDepartures.selectLoading
);
