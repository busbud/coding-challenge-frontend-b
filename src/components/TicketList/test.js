import React from 'react';
import renderer from 'react-test-renderer';

import { TicketList } from './';
import { defaultState } from '../../reducers/';

const noOp = () => '';

describe('Component', () => {
	it('should render', () => {
		const wrapper = renderer.create(
			<TicketList t={noOp} tickets={defaultState.tickets} />
		).toJSON();
		expect(wrapper).toMatchSnapshot();
	});
});
