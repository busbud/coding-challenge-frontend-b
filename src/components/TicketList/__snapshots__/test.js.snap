// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component should render 1`] = `
<section
  className="ticket-list"
>
  <article
    className="ticket"
  >
    <header
      className="heading header"
    >
      <div
        className="columns"
      >
        <div
          className="column"
        >
          <section>
            
            : 
            <strong>
              10:00 AM
            </strong>
          </section>
        </div>
        <div
          className="column"
        >
          <section>
            
            : 
            <strong>
              1:35 AM
               
              (+1)
            </strong>
          </section>
        </div>
        <div
          className="column"
        >
          <section>
            
            : 
            <strong>
              
              		15  35 
              	
            </strong>
          </section>
        </div>
      </div>
    </header>
    <div
      className="columns"
    >
      <div
        className="column"
      >
        <section
          className="partner"
        >
          <a
            className="link"
            onClick={undefined}
            rel="noopener noreferrer"
            target="_blank"
          >
            <header
              className="logo"
            >
              <img
                alt={undefined}
                src={undefined}
              />
            </header>
            <p />
          </a>
        </section>
      </div>
      <div
        className="column"
      >
        <section
          className="details"
        >
          <p>
            <a
              className="link"
              onClick={[Function]}
            >
              NY
            </a>
             â†’ 
            <a
              className="link"
              onClick={[Function]}
            >
              MR
            </a>
          </p>
          <p>
            <strong>
              
            </strong>
             
            
             
            <strong>
              15
               
              
               
              35
               
              
            </strong>
          </p>
          <p>
            
            : 
            
          </p>
        </section>
      </div>
      <div
        className="column"
      >
        <section
          className="buy"
        >
          <p>
            $95.43
          </p>
          <button
            className="button"
            onClick={[Function]}
          >
            
          </button>
        </section>
      </div>
    </div>
  </article>
</section>
`;
